crates/editor/src/editor.rs

```diff
@@ -3995,6 +4013,127 @@ impl Editor {
         }))
     }
 
+    fn hide_signature_help(&mut self, cx: &mut ViewContext<Self>) -> bool {
+        if self.signature_help_state.is_some() {
+            self.signature_help_state = None;
+            self.signature_help_task = None;
+            cx.notify();
+            true
+        } else {
+            false
+        }
+    }
+
+    fn should_open_signature_help_automatically(
+        &mut self,
+        old_cursor_position: &Anchor,
+        cx: &mut ViewContext<Self>,
+    ) -> bool {
+        if !EditorSettings::get_global(cx).auto_signature_help {
+            return false;
+        }
+        let newest_selection = self.selections.newest::<usize>(cx);
+        let head = newest_selection.head();
+        if !newest_selection.is_empty() || head != newest_selection.tail() {
+            self.signature_help_state = None;
+            self.signature_help_hidden_by_selection = head != newest_selection.tail();
+            return false;
+        }
+
+        let buffer_snapshot = self.buffer().read(cx).snapshot(cx);
+        let previous_position = old_cursor_position.to_offset(&buffer_snapshot);
+        let previous_brackets_surround = buffer_snapshot
+            .innermost_enclosing_bracket_ranges(previous_position..previous_position, None)
+            .filter(|(start_bracket_range, end_bracket_range)| {
+                start_bracket_range.start != previous_position
+                    && end_bracket_range.end != previous_position
+            });
+        let current_brackets_surround = buffer_snapshot
+            .innermost_enclosing_bracket_ranges(head..head, None)
+            .filter(|(start_bracket_range, end_bracket_range)| {
+                start_bracket_range.start != head && end_bracket_range.end != head
+            });
+        match (previous_brackets_surround, current_brackets_surround) {
+            (None, None) => {
+                self.signature_help_state = None;
+                false
+            }
+            (Some(_), None) => {
+                self.signature_help_state = None;
+                false
+            }
+            (None, Some(_)) => {
+                self.signature_help_hidden_by_selection = false; // Reset the flag
+                true
+            }
+            (Some(previous), Some(current)) => {
+                if self.signature_help_hidden_by_selection {
```

Comment:
I believe there's a flaw in the logic, or at least the behavior is somewhat confusing: if I click between brackets and get a popover, I seem to be able to ESC it away and it never appears again until I navigate away from brackets and come back — this is neat.

But if, after that ESC to close the popover, I do select & unselect staying within the brackets, I get the popover back.
Should we add a test and fix that?

Reply:
ESC behavior is almost there, but one more case is still on: the same `("` case from the video above now restores the popover when the caret is between `(` and `"`.

Reply:
@SomeoneToIgnore 
Thank you. It seems that we can solve this by defining it with reference to the quotes-related definitions in surrounds.rs and passing it to range_filter, but this would involve hardcoding the characters. I would like to hear your opinion on this approach. If there is a problem, I will revert it, but it seems that there are no other good methods.

fixed it on 18f4b56b9

---

```rust
let range_filter = |start: Range<usize>, end: Range<usize>| {
    let text = buffer_snapshot.text();
    let (text_start, text_end) = (text.get(start), text.get(end));
    QUOTES_PAIRS
        .into_iter()
        .all(|(start, end)| text_start != Some(start) && text_end != Some(end))
};

let previous_position = old_cursor_position.to_offset(&buffer_snapshot);
let previous_brackets_range = bracket_range(previous_position);
let previous_brackets_surround = buffer_snapshot
    .innermost_enclosing_bracket_ranges(previous_brackets_range, Some(&range_filter))
    .filter(|(start_bracket_range, end_bracket_range)| {
        start_bracket_range.start != previous_position
            && end_bracket_range.end != previous_position
    });
let current_brackets_range = bracket_range(head);
let current_brackets_surround = buffer_snapshot
    .innermost_enclosing_bracket_ranges(current_brackets_range, Some(&range_filter))
    .filter(|(start_bracket_range, end_bracket_range)| {
        start_bracket_range.start != head && end_bracket_range.end != head
    });
```

### The definition in surrounds.rs that I referred to.
```rust
Object::Quotes => Some(BracketPair {
    start: "'".to_string(),
    end: "'".to_string(),
    close: true,
    surround: true,
    newline: false,
}),
Object::BackQuotes => Some(BracketPair {
    start: "`".to_string(),
    end: "`".to_string(),
    close: true,
    surround: true,
    newline: false,
}),
Object::DoubleQuotes => Some(BracketPair {
    start: "\"".to_string(),
    end: "\"".to_string(),
    close: true,
    surround: true,
    newline: false,
}),
```

### Movie
https://github.com/zed-industries/zed/assets/55743826/6387d7d8-43f2-4fda-9d2c-b74665022164